### PR TITLE
feat(stepfunctions-tasks): add httpinvoke step functions task

### LINKED ISSUE #28278: aws-stepfunctions-tasks: Support for new HTTPS Endpoint integration
### Describe the feature

Support for the newly released HTTPS Endpoints Step Functions integration:
- https://aws.amazon.com/about-aws/whats-new/2023/11/aws-step-functions-https-endpoints-teststate-api/
- https://aws.amazon.com/blogs/aws/external-endpoints-and-testing-of-task-states-now-available-in-aws-step-functions/

### Use Case

CDK should support the HttpInvoke Step Functions integration, so it can be managed via IaC.

### Proposed Solution

Task should allow specifying all standard, optional, and required parameters (endpoint, HTTP method, EventBridge Connection ARN, headers, body, query string, etc.).

### Other Information

_No response_

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

latest

### Environment details (OS name and version, etc.)

N/A

### PR DESCRIPTION
**Pull request by MathieuGilbert**
This adds an `HttpInvoke` Step Functions task construct, which allows calling public APIs as described [here](https://docs.aws.amazon.com/step-functions/latest/dg/connect-third-party-apis.html).

- Accepts an Event Bridge API Destination `IConnection`, setting required permissions to use its credentials.
- Grants permission to invoke the root path for the endpoint (`apiRoot` prop). This allows passing the relative endpoint path at execution time (`apiEndpoint` prop).
- Defines an enum for the allowed options for the URL encoding style of arrays, used when encoding the request body.

Closes #28278

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/README.md [hunk 1]
```diff
@@ -1066,6 +1066,32 @@ new tasks.GlueDataBrewStartJobRun(this, 'Task', {
 });
 ```
 
+## Invoke HTTP API
+
+Step Functions supports [calling third-party APIs](https://docs.aws.amazon.com/step-functions/latest/dg/connect-third-party-apis.html) with credentials managed by Amazon EventBridge [Connections](https://docs.aws.amazon.com/eventbridge/latest/APIReference/API_Connection.html).
+
+The following snippet creates a new API destination Connection, and uses it to make a POST request to the specified URL. The endpoint response is available at the `$.ResponseBody` path.
```
Comment (msambol) [inline packages/aws-cdk-lib/aws-stepfunctions-tasks/README.md:1073]: nit: `Connection`

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/http/invoke.ts [hunk 2]
```diff
@@ -0,0 +1,180 @@
+import { Construct } from 'constructs';
+import * as events from '../../../aws-events';
+import * as iam from '../../../aws-iam';
+import * as sfn from '../../../aws-stepfunctions';
+import { integrationResourceArn } from '../private/task-utils';
+
+/**
+ * The style used when applying URL encoding to array values.
+ *
```
Comment (msambol) [inline packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/http/invoke.ts:9]: nit: extra newline

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/http/invoke.ts [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+  /**
+   * The API endpoint to call, relative to `apiRoot`.
+   * @example sfn.TaskInput.fromText("path/to/resource")
+   *
+   */
+  readonly apiEndpoint: sfn.TaskInput;
+
+  /**
+   * The HTTP method to use.
+   * @example sfn.TaskInput.fromText("GET")
```
Comment (msambol) [inline packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/http/invoke.ts:50]: nit: can you add a newline before `@example` and remove the one below.

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/http/invoke.ts [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+  /**
+   * Permissions are granted to call all resources under this path.
+   * @example "https://api.example.com"
+   *
+   */
+  readonly apiRoot: string;
+
+  /**
+   * The API endpoint to call, relative to `apiRoot`.
+   * @example sfn.TaskInput.fromText("path/to/resource")
```
Comment (msambol) [inline packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/http/invoke.ts:43]: likewise here.

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/http/invoke.ts [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+  /**
+   * The HTTP method to use.
+   * @example sfn.TaskInput.fromText("GET")
+   *
+   */
+  readonly method: sfn.TaskInput;
+
+  /**
+   * The EventBridge Connection to use for authentication.
+   *
```
Comment (msambol) [inline packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/http/invoke.ts:57]: nit: extra newline

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/http/invoke.ts [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+   * @default - No body is sent with the request.
+   */
+  readonly body?: sfn.TaskInput;
+
+  /**
+   * The headers to send to the HTTP endpoint.
+   *
+   * @default - No additional headers are added to the request.
+   * @example sfn.TaskInput.fromObject({ 'Content-Type': 'application/json' })
```
Comment (msambol) [inline packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/http/invoke.ts:72]: nit: can you put `@default` below `@example` and add a newline in between.

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/http/invoke.ts [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+   * Only used if `urlEncodeBody` is `true`.
+   *
+   * @default - ArrayEncodingFormat.INDICES
+   */
+  readonly arrayEncodingFormat?: ArrayEncodingFormat;
+}
+
+/**
+ * A Step Functions Task to call a public third-party API.
+ *
```
Comment (msambol) [inline packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/http/invoke.ts:103]: nit: extra newline.

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/http/invoke.ts [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+ */
+export class HttpInvoke extends sfn.TaskStateBase {
+  protected readonly taskMetrics?: sfn.TaskMetricsConfig;
+  protected readonly taskPolicies?: iam.PolicyStatement[];
+
+  constructor(
+    scope: Construct,
+    id: string,
+    private readonly props: HttpInvokeProps,
```
Comment (msambol) [inline packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/http/invoke.ts:112]: nit: Can these be on the same line?

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/http/invoke.ts [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+
+    this.taskPolicies = this.buildTaskPolicyStatements();
+  }
+
+  /**
+   * Provides the HTTP Invoke service integration task configuration.
+  */
+  /**
+   * @internal
+  */
```
Comment (msambol) [inline packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/http/invoke.ts:124]: nit: this formatting looks off

**Review by msambol — CHANGES_REQUESTED** [review]
Great stuff. Mostly nits on formatting.

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/http/invoke.ts [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+   */
+  readonly urlEncodeBody?: boolean;
+
+  /**
+   * The format of the array encoding.
+   * Only used if `urlEncodeBody` is `true`.
+   *
+   * @default - ArrayEncodingFormat.INDICES
+   */
+  readonly arrayEncodingFormat?: ArrayEncodingFormat;
```
Comment (TheRealAmazonKendra) [inline packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/http/invoke.ts:98]: Given that this field is reliant on `urlEncodeBody` being set, these two fields should probably be one field. Maybe the only field here should still be called `urlEncodeBody` but have it take an `Encoding` (not sure about this name) class with a static function that has an optional input of `ArrayEncodingFormat`. Something along the lines of 
```
urlEncodeBody: Encoding.set(ArrayEncodingFormat.INDICES);
```
to include encoding format, or 
```
urlEncodeBody: Encoding.set();
```
to just include encoding with the default format.

This way, if they leave off this prop altogether, it's defaulted to `false`.

Comment (TheRealAmazonKendra) [reply packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/http/invoke.ts:98]: I'll also add the note here that `set` probably isn't a great function name here.

**Review by TheRealAmazonKendra — DISMISSED** [review]
Thank you for your contribution! This looks great overall but I just have one suggested change (in addition to the feedback already given).

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/http/invoke.ts (hunk 10 above)
Comment (MathieuGilbert) [reply packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/http/invoke.ts:98]: Thanks for pointing this out, I agree it's a bit awkward with the 2 params. I played around a bit, and I think a single prop can work with a couple extra enum options and renaming it to `urlEncodingFormat`. You can not pass it or specify `URLEncodingFormat.NONE` to get no encoding, pass it as `URLEncodingFormat.DEFAULT` to get encoding with the default array encoding (the underlying service defaults to `INDICES`), or specify the existing values to use those.

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/http/invoke.ts (hunk 9 above)
Comment (MathieuGilbert) [reply packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/http/invoke.ts:124]: Thanks for pointing these out, I obviously didn't pay enough attention to the comments (and don't normally use them). I did a sweep and they should all be consistent now.

**Comment by MathieuGilbert** [comment]
@msambol @TheRealAmazonKendra Thanks for reviewing, I think I've addressed all of your comments. Please let me know if there's anything else!

**Review by TheRealAmazonKendra — APPROVED** [review]
Thank you for all your work on this!

**Comment by TheRealAmazonKendra** [comment]
@Mergifyio update

**Review by TheRealAmazonKendra — APPROVED** [review]

**Pull request merged**